process.go

```diff
@@ -0,0 +1,183 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"io"
+
+	"errors"
+)
+
+// == Direct input processing ========================================
+
+type renderResult struct {
+	err     error
+	inPath  string
+	outPath string
+}
+
+func processInputFiles(stringTemplate string, inFiles []string, outFiles []string, g *Gomplate) error {
+	inFiles, err := readInputs(stringTemplate, inFiles)
+	if err != nil {
+		return err
+	}
+
+	if len(outFiles) == 0 {
+		outFiles = []string{"-"}
+	}
+
+	results := make(chan *renderResult)
+	defer close(results)
+
+	for n, input := range inFiles {
+		go func(idx int, input string) {
+			err := renderTemplate(g, input, outFiles[idx])
+			results <- &renderResult{err, input, outFiles[idx]}
+		}(n, input)
+	}
+
+	return waitAndEvaluateResults(results, len(inFiles))
+}
+
+// == Recursive input dir processing ======================================
+
+func processInputDir(inputDir string, outDir string, g *Gomplate) error {
+	results := make(chan *renderResult)
+	defer close(results)
+	nr := processDir(inputDir, outDir, g, results, 0)
+	return waitAndEvaluateResults(results, nr)
+}
+
+func processDir(inPath string, outPath string, g *Gomplate, results chan *renderResult, nr int) int {
+	inPath = filepath.Clean(inPath)
+	outPath = filepath.Clean(outPath)
+
+	// assert tha input path exists
+	si, err := assertDir(inPath)
```

Comment:
From the documentation its not clear to me whether ReadDir fails properly (e.g. with a proper error message). Also the file status `si`is needed so that the output dir will have the same file mode (btw, currently the generated files all have fixed 0644, but shouldn't the have the same mode as the original template ?)

I could inline the method since it is used now only once. Will rename it to `statDir` at least.